Sensor base 0.7.0

jeffmahoney opened the pull request:
This PR updates the base version of Velociraptor to 0.7.0 and re-applies the Linux Security Sensor extensions.

jeffmahoney committed:
Add parser to read systemd journal on Linux

Newer Linux systems use the systemd journal facility for logging instead
of the classic syslogd logging daemon.  While it's possible to configure
journald to forward events to e.g. rsyslogd, doing so requires a
configuration change on the endpoint that would only be necessary to
accommodate the syslog parser.

This commit adds the ability to parse the systemd journal using the
native libsystem-journal.so shared library.  It adds watcher and scanner
parsers identical to the syslog versions.  One difference is that since
the underlying library handles opening and tracking the journal files,
there is no facility to subscribe to specific journal files.  The
support is also limited to the system journal, but that should
be sufficient for Velociraptor.  It is expected that queries against
these parsers filter on systemd unit (or other fields)
as desired rather than implementing those filters in the parser itself.

jeffmahoney committed:
Update the Linux.Syslog.SSHLogin artifact to scan the systemd journal

This adds an additional query to the Linux.Syslog.SSHLogin artifact to
iterate over the systemd journal.  The default systemd unit is
sshd.service but it can be configured differently.  The Grok query
is reused, making the syslog-specific portions of the string optional.

jeffmahoney committed:
Update the Linux.Events.SSHLogin artifact to scan the systemd journal

This adds an additional query to the Linux.Events.SSHLogin artifact to
iterate over the systemd journal.  The default systemd unit is
sshd.service but it can be configured differently.  The Grok query
is reused, making the syslog-specific portions of the string optional.

This also adds a MONITORING_DAILY report since the default query only
shows the _first_ 100 events, which is likely not what anybody wants.

lorddoskias committed:
linux: add lsattr() function to enumerate file attributes

Originally by Nikolay Borisov <nborisov@suse.com>

lorddoskias committed:
Linux.Detection.ImmutableFiles: Enumerate immutable files under a path

jeffmahoney committed:
Add systemd-dev as build dependency for github workflow

With the journald code added, systemd-dev is a build dependency.

jeffmahoney committed:
SUSE: Do build tests on every pull request

Our primary target is Linux.  We want to ensure that a PR builds on Linux
before it hits a real branch.

lorddoskias committed:
vql/linux/bpf/tcpsnoop: Add tcpsnoop plugin

The plugin utilizes the ebpf infrastructure to load a small program
which hooks into tcp_connect/tcp_accept code paths of the kernel and
print information about the connection source/dest ports/ip addresses,
PID/UID of the process initiating the connection, ip protocol version,
timestamp, and the direction of the connection.

The Linux networking subsystem supports an xt_AUDIT target that, on the
surface, looks like it would serve a similar purpose but it provides
no information about user or pid.

Co-authored-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
SUSE: Add SSHLogin artifacts

The SSHLogin artifacts as shipped with Velociraptor are inconsistent
with one another.  These artifacts resolve the inconsistencies without
breaking existing VRR installs.

guangyee committed:
Add custom artifacts for login and logout attempts recorded by auditd

chajain committed:
Linux.Events.ProcessExecutions: catch 32-bit execve calls

lorddoskias committed:
vql/linux/bpf/dnssnoop: Add dnssnoop() plugin

It works by creating a raw socket and attaching an ebpf program to
filter only incoming dns packets. That way replies can be parsed.

Signed-off-by: Nikolay Borisov <nborisov@suse.com>
Co-authored-by: Jeff Mahoney <jeffm@suse.com>

lorddoskias committed:
vql/linux/bpf/chattrsnoop: Add plugin to catch changes to inode attributes

It hooks into do_vfs_ioctl and also performs some of the checks that the
generic code would perform in order to ensure we are triggering the ebpf
kprobe on a fs which indeed supports the given operations.

While at it remove the vmlinux.h files that comes with tcpsnoop and
instead rely on the one provided by libbpf.

Co-authored-by: Jeff Mahoney <jeffm@suse.com>

guangyee committed:
Add artifact to monitor user group updates (#24)

Add artifact collects user group changes from auditd events, by
monitoring updates made to either /etc/group or /etc/nsswitch.conf.

chajain committed:
Extend audit artifacts to use new interface

Now that the audit plugin monitors the rules, we can specify the
required rules in the plugin call instead of calling auditctl.

jeffmahoney committed:
vql/linux/audit: use go-libaudit v2 for live audit message processing

In 2020, the upstream go-libaudit project moved to v2, changing the
module path to github.com/elastic/go-libaudit/v2.

Upcoming changes to the audit plugin to improve scalability and allow
individual rule deletion require some changes that are pending upstream
acceptance:

- Reporting rules that were missing when the rule is deleted
- Not assuming that no syscalls means all syscalls when building rules
- Support for removing individual rules
- Adding a new AuditMessage.Parse method to allow the caller to manage
  the lifecycle of the AuditMessage
- Message lifecycle

For now, we have a replace rule in effect that points to the
github.com/jeffmahoney/go-libaudit/v2 repository until those changes
land.

Side effects: watch_auditd and parse_auditd must now be Linux-only due
to the use of unix.SignalName in go-libaudit/v2.  I doubt this will have
much of an impact since parsing Linux kernel audit logs in an artifact
on non-Linux systems seems pretty unlikely.

jeffmahoney committed:
vql/linux/audit: rearchitect plugin for scalability

The audit plugin doesn't share anything among invocations, which makes
sense given the historical model of preconfiguring of the rules,
subscribing to the firehose, and sorting out the events later.  The problem
with that architecture is that if anything external happens with the
audit subsystem -- even as simple as restarting the audit daemon -- the
audit rules are lost and no more events will be generated.

By specifying which rules a particular artifact depends on in the
arguments, we can be certain that the rules required will always be in
effect, can log when they are not, and restore them automatically.

Since multiple invocations can ultimately require the same audit rules,
management of those rules must be centralized and reference counted so
they are only removed when the last plugin using a rule exits.

There are a few changes required to make this go:
- We need a separate command-and-control audit client to do the
  configuration, checking, and fixing
- We need a goroutine that periodically checks to ensure that the rules
  are correct.  This is implemented as something that sits idle waiting
  for configuration events and then batches them until 500ms has passed
  since receipt of the last one.  The reason for this is that e.g.
  "auditctl -D" will generate an event for every single rule removed
  and we don't need to check the rules every time.
- Each invocation will receive a copy of each event and also log
  messages that apply to the invocation as needed.
- The lifetime of the shared services is tied to the listeners.
  Once the last listener exits, the shared services shutdown.  If new
  invocations come in, a new instance will be started up.
- In order to not conflict with auditd, we must use the multicast
  interface for receiving messages.  The kernel does *not* queue and
  retry multicast audit messages so we need to be very careful to avoid
  dropping events.
- In order to avoid dropping events on very busy systems, we increase
  the default socket receive buffer to 512k and allow it to scale up
  as needed in response to ENOBUFS from the kernel.
- We use a separate buffer for each event received so that the
  reception of the messages and processing of events can be decoupled.
- These buffers are kept in a pool to avoid constant allocation and
  freeing.
- We use a FileBytesRingBuffer to queue events so that we don't end up pinning
  lots of memory if events get backlogged.
- To aid in debugging and monitoring, there are statistics gathered.  A
  global variable can be used to enable printing the statistics to the
  client terminal every 5 seconds.

chajain committed:
Updating the NewFiles and ProcessStatuses Artifacts

lorddoskias committed:
vql/linux/cronsnoop: Add cronsnoop() plugin

The cronsnoop plugin uses fsnotify to create events when cron jobs
are added/removed/changed.

Co-authored-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
Add Linux.Sys.Bash to Server.Monitor.Shell artifact

This needs to be a separate source since the Windows shell artifacts
and Linux.Sys.Bash present the same information in different ways.

jeffmahoney committed:
SUSE: add Docker files

jeffmahoney committed:
SUSE: Add docker-compose environment

jeffmahoney committed:
SSHLogin: require _TRANSPORT != 'kernel' from watch_journal()

When events are received by the journal from the kernel they do not
have a _SYSTEMD_UNIT key.  This results in noisy error messages in the
log.  We can avoid this by requiring that _TRANSPORT != 'kernel'.

jeffmahoney committed:
vql/linux/bpf: add support to add bpf plugins for Linux

We're about to add several plugins that operate using BPF.  This package
adds some helpers that will be shared among them.

We use the libbpfgo module to implement the BPF interface.  The libbpfgo
project wants to build its own libbpf and statically link with it.  In order
to build the embedded libbpf, we need to incorporate libbpfgo as a submodule.

There are external requirements to build the BPF programs:
- bpftool (unless the builder provides a pregenerated vmlinux.h)
- clang >= 10
- llvm-strip

Since these external requirements are new, the BPF plugins will only be
enabled if the above requirements are met.  If they are unmet, a message will be
printed informing the user about the missing tools and the build will
continue.  The plugins can be made mandatory by setting the environment
variable: BUILD_BPF_PLUGINS=1.

Cross-builds of BPF modules are unsupported with this iteration.

Co-authored-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
Github: Run build workflow on each pull request

jeffmahoney committed:
github: run testcases on Linux builds in new workflow

We don't need to build the GUI or the binary on every architecture before
running the testcases.

jeffmahoney committed:
rpm: introduce rpm vql plugin

This commit adds the rpm() vql plugin that will return all installed
RPM packages on the system.  Tags can be specified using the tags
parameter.  Alternate paths to the directory containing the database files can
be specified using the paths parameter.

jeffmahoney committed:
magefile.go: use current architecture for Linux builds

When building in a multiarchitecture build system, it's helpful to have
the same commands for every architecture.

This commit makes the Linux and LinuxBare targets default to the host
architecture and creates new LinuxAmd64 and LinuxBareAmd64 targets to
target that architecture specifically.

jeffmahoney committed:
vql/functions/hash: cache results on Linux

When we use Velociraptor on systems with a big fork/exec workload,
use system call auditing, and hash the executable, we end up with
pinned CPUs for obvious reasons.  There's no value in hashing the same
file over and over if we know it hasn't changed.

This commit adds an LRU cache for the hash results.  It uses the
dev number, inode number, mtime, and ctime to ensure that the file
hasn't been modified.  A race does exist betewen stat and open, but
for this use case a bigger race exists between exec and the hash
being requested.

jeffmahoney committed:
github/workflows/linux: do apt-get update to refresh package lists

jeffmahoney committed:
file_store/directory/buffer: add direct-serialized interface

The file-based ring buffer implementation works well for generic data
buffering, but there's no need to take a byte stream, turn it into an
ordered dict and then marshal it to json to use it.

This commit adds EnqueueSerialized and LeaseSerialized methods to allow
direct queuing of bytestreams.

jeffmahoney committed:
file_store/directory/listener_bytes: Add listener to use serialized interface

The previous commit allowed the FileBasedRingBuffer to use direct byte
sequences.  This adds the listener to use it.

scudette committed:
Update the config file docs.

snyk-bot committed:
fix: upgrade recharts from 2.7.2 to 2.7.3

Snyk has created this PR to upgrade recharts from 2.7.2 to 2.7.3.

See this package in npm:
https://www.npmjs.com/package/recharts

See this project in Snyk:
https://app.snyk.io/org/scudette/project/76f4d127-566b-42ef-86f4-bdcbc92b90b4?utm_source=github&utm_medium=referral&page=upgrade-pr

scudette committed:
[Snyk] Upgrade: @babel/core, @babel/runtime (#2917)

scudette committed:
Bugfixes: Artifact bugs due to FullPath->OSPath refactor (#2923)

misje committed:
Fix AptSources deb822 parsing bug and add deb822 test  (#2926)

#2851 was merged a bit too early. My mistake, I should have marked it as
a draft. A misplaced `WHERE URIs` prevented deb822 parsing to work
altogether. This is now fixed, and I've included some sources files to
the tests, along with another lists file that includes two common
options.

I've added these new files to the tests Sources and SourcesFlattened,
but not SourcesCacheFiles.

I've also fixed the VQL suggestions after realising that `get()` doesn't
work in WHERE as I assumed it would (see discussion
[here](https://discord.com/channels/624244632552734750/865024982399320114/1147470384359284746)
with a detailed explanation from Mike).

scudette committed:
[Snyk] Upgrade ace-builds from 1.23.4 to 1.24.0 (#2928)

Zawadidone committed:
Remove unused parameter by Bloodhound artifact (#2924)

The executable is deleted by default.

```
[...]
[INFO] 2023-09-01T09:46:13-07:00 copy: Copying file from /uploads/SharpHound.exe into C:\Users\IEUser\AppData\Local\Temp\tmp905568840.exe
[INFO] 2023-09-01T09:46:13-07:00 Starting collection of Windows.ActiveDirectory.BloodHound
[INFO] 2023-09-01T09:46:13-07:00 shell: Running external command [C:\Users\IEUser\AppData\Local\Temp\tmp905568840.exe --outputdirectory C:\Users\IEUser\AppData\Local\Temp\tmp1873669653 --nozip --outputprefix MSEDGEWIN10.localdomain]
[...]
[INFO] 2023-09-01T09:46:37-07:00 tempfile: removing tempfile C:\Users\IEUser\AppData\Local\Temp\tmp905568840.exe
```

scudette committed:
Add Modify() method to client info manager. (#2933)

This is required to ensure client info records can be modified
atomically. Fixes certain edge cases arising due to server monitoring
queries assigning labels racing with interrogations.

scudette committed:
[Snyk] Upgrade ace-builds from 1.24.0 to 1.24.1 (#2932)

<p>This PR was automatically created by Snyk using the credentials of a
real user.</p><br /><h3>Snyk has created this PR to upgrade ace-builds
from 1.24.0 to 1.24.1.</h3>

:information_source: Keep your dependencies up-to-date. This makes it
easier to fix existing vulnerabilities and to more quickly identify and
fix newly disclosed vulnerabilities when they affect your project.
<hr/>

- The recommended version is **1 version** ahead of your current
version.
- The recommended version was released **24 days ago**, on 2023-08-15.


<details>
<summary><b>Release notes</b></summary>
<br/>
  <details>
    <summary>Package name: <b>ace-builds</b></summary>
    <ul>
      <li>
<b>1.24.1</b> - <a
href="https://snyk.io/redirect/github/ajaxorg/ace-builds/releases/tag/v1.24.1">2023-08-15</a></br><p>package
15.08.23</p>
      </li>
      <li>
<b>1.24.0</b> - <a
href="https://snyk.io/redirect/github/ajaxorg/ace-builds/releases/tag/v1.24.0">2023-08-09</a></br><p>package
09.08.23</p>
      </li>
    </ul>
from <a
href="https://snyk.io/redirect/github/ajaxorg/ace-builds/releases">ace-builds
GitHub release notes</a>
  </details>
</details>


<details>
  <summary><b>Commit messages</b></summary>
  </br>
  <details>
    <summary>Package name: <b>ace-builds</b></summary>
    <ul>
<li><a
href="https://snyk.io/redirect/github/ajaxorg/ace-builds/commit/d065fe88b942824db5e8820b8e9693df3499a2cb">d065fe8</a>
package 15.08.23</li>
    </ul>

<a
href="https://snyk.io/redirect/github/ajaxorg/ace-builds/compare/cd3ae1b04ed894bad5306210ce50796aa6556468...d065fe88b942824db5e8820b8e9693df3499a2cb">Compare</a>
  </details>
</details>
<hr/>

**Note:** *You are seeing this because you or someone else with access
to this repository has authorized Snyk to open upgrade PRs.*

For more information: <img
src="https://api.segment.io/v1/pixel/track?data=eyJ3cml0ZUtleSI6InJyWmxZcEdHY2RyTHZsb0lYd0dUcVg4WkFRTnNCOUEwIiwiYW5vbnltb3VzSWQiOiIwMTExMWI4MC01NWU4LTQwYmQtOWU3Ni1kNTFhMDE2MGIwZTciLCJldmVudCI6IlBSIHZpZXdlZCIsInByb3BlcnRpZXMiOnsicHJJZCI6IjAxMTExYjgwLTU1ZTgtNDBiZC05ZTc2LWQ1MWEwMTYwYjBlNyJ9fQ=="
width="0" height="0"/>

🧐 [View latest project
report](https://app.snyk.io/org/scudette/project/76f4d127-566b-42ef-86f4-bdcbc92b90b4?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;upgrade-pr)

🛠 [Adjust upgrade PR
settings](https://app.snyk.io/org/scudette/project/76f4d127-566b-42ef-86f4-bdcbc92b90b4/settings/integration?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;upgrade-pr)

🔕 [Ignore this dependency or unsubscribe from future upgrade
PRs](https://app.snyk.io/org/scudette/project/76f4d127-566b-42ef-86f4-bdcbc92b90b4/settings/integration?pkg&#x3D;ace-builds&amp;utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;upgrade-pr#auto-dep-upgrades)

<!---
(snyk:metadata:{"prId":"01111b80-55e8-40bd-9e76-d51a0160b0e7","prPublicId":"01111b80-55e8-40bd-9e76-d51a0160b0e7","dependencies":[{"name":"ace-builds","from":"1.24.0","to":"1.24.1"}],"packageManager":"npm","type":"auto","projectUrl":"https://app.snyk.io/org/scudette/project/76f4d127-566b-42ef-86f4-bdcbc92b90b4?utm_source=github&utm_medium=referral&page=upgrade-pr","projectPublicId":"76f4d127-566b-42ef-86f4-bdcbc92b90b4","env":"prod","prType":"upgrade","vulns":[],"issuesToFix":[],"upgrade":[],"upgradeInfo":{"versionsDiff":1,"publishedDate":"2023-08-15T11:47:56.988Z"},"templateVariants":[],"hasFixes":false,"isMajorUpgrade":false,"isBreakingChange":false,"priorityScoreList":[]})
--->

Co-authored-by: snyk-bot <snyk-bot@snyk.io>

scudette committed:
Send hard coded labels in periodic client info updates (#2935)

It is possible to hard code labels into the config file. These labels
will be automatically added when an interrogation process occurs

This PR also includes them in periodic client info updates

scudette committed:
Report process names as well as pid for errors (#2937)

In Windows many processes are protected so we can not access them.
Previously Velociraptor only reported the pid but this PR also reports
the process name for inaccessible processes

misje committed:
Add a generic hostsfile artifact (#2930)

I wrote an artifact for parsing hosts files on Linux/macOS. Afterwards,
I realised there already [exists one for
Windows](https://github.com/Velocidex/velociraptor/blob/master/artifacts/definitions/Windows/System/HostsFile.yaml).
The file format is the same on all three supported OSes, so I don't see
a reason for not having a generic artifact that runs on all of them.

The differences now are

- I match the array of hostnames against the regex, allowing one to use
anchors like "^"/"$" in the regex (instead of matching the string of
hostname/aliases)
- I provide flattened query as well, with one hostname per row
- The parsing function is exported
- The hosts filename parameter is now a glob

Let me know what you want to do with this. Include it and leave the
existing Windows.System.HostsFile? Replace the existing? Suggest this to
belong in the exchange instead? Something else?

I'll add the test when I know how to proceed.

---------

Co-authored-by: Mike Cohen <mike@velocidex.com>

scudette committed:
Order columns in hostname flatten output (#2939)

jeffmahoney committed:
build: update to mage 0.15

This allows using a magefiles/ directory for target definitions.

jeffmahoney committed:
Users testcases (#2942)

This PR extends the testing around user state.

---------

Co-authored-by: Mike Cohen <mike@velocidex.com>

jeffmahoney committed:
Update bluemonday dependency. (#2941)

The github CI runs fail with:

go: warning: github.com/microcosm-cc/bluemonday@v1.0.21: retracted by
module author: Retract older versions as only latest is to be depended
upon go: to switch to the latest unretracted version, run:
	go get github.com/microcosm-cc/bluemonday@latest

hillu committed:
Improve various bits of VQL documentation (#2945)

Co-authored-by: Mike Cohen <mike@velocidex.com>

jeffmahoney committed:
vql/linux/bpflib: add sample vmlinux.h includes for test builds

The test builds use Ubuntu containers and Ubuntu tries _very_ hard to
match the kernel-related tools with the kernel version.  It's much
easier to just provide vmlinux.h.

scudette committed:
[Snyk] Upgrade: @babel/core, @babel/runtime (#2948)

djoreilly commented on vql/linux/audit/audit_service.go:
```diff
@@ -0,0 +1,1018 @@
+			if err != nil {
+				self.logger.Info("failed to parse message: %v", err)
+			}
+			self.rawBufPool.Put(buf)
```
testing this branch with
```
velociraptor-v0.7.0-linux-amd64 -v query "select * from audit(rules=\"-a always,exit -F arch=b64 -S execve -k vrr_procmon\")" 
````
and running commands to trigger events, i'm finding most events are empty (fields are not filled in). The first event is always okay, but the next ones are likely to be empty. I think it's something to do with reusing buffers from the pool, because the events have all their fields filled if I comment out this line.

jeffmahoney replied:
Yep, that was definitely it. I now have test cases to ensure that the data output is sane.

scudette committed:
Relay hunt creation errors to the Hunts API (#2953)

During refactor of the CreateHunt API to be able to support multiple
orgs the errors from launching the hunt on any org were reported in the
log and not relayed to the caller.

This means that the errors were not propagated to the API so if the
artifact failed to compile within the specified org no error was visible
to the user.

This PR collects all errors and reports them together therefore ensuring
the user knows the hunt can not be launched.

FIXES: #2952

djoreilly commented on vql/linux/audit/audit_service.go:
```diff
@@ -0,0 +1,1018 @@
+			continue
+		}
+
+		self.messageQueue.Send(buf[:size])
```
I think this needs to be `self.messageQueue.Send(buf)` so Put() in queuemainEventLoop() will put the whole buffer back in the pool. 
https://gist.github.com/jaysoncena/f2c98cba63f26a181c0ebb216b80d3de#file-bytepool1-go-L20-L24

djoreilly replied:
I spent some more time looking at this. Sending to messageQueue means it might goes to disk, and I think we read back from disk into some other address, and that should not be put back into the pool.

I found this fixes the problem
```
               bufCopy := make([]byte, size)
               copy(bufCopy, buf)
               self.rawBufPool.Put(buf)
               self.messageQueue.Send(bufCopy)
```
and remove `self.rawBufPool.Put(buf)` from `mainEventLoop()`. 

But maybe this defeats the benefit of using sync.Pool in the first place.

jeffmahoney replied:
I was thinking about this as well and I think it's clear it's not fully baked. I think the sync pool still makes sense, but the lifecycle needs to be fixed. I've mostly finished up reworking the bpf stuff to be (hopefully) acceptable for submission and will take another look at this.

jeffmahoney replied:
This one has been resolved.

hillu committed:
Fix device major/minor number calculations (#2958)

So far, Velociraptor seemed to assume that Unix device numbers were
evenly split 16 bit numbers, this hasn't been true in decades.

This is probably most important on Linux since there are a lot of
published artifacts that treat device numbers as magic indicators for
real/local filesystems.

scudette committed:
[Snyk] Upgrade axios from 1.4.0 to 1.5.0 (#2951)

scudette committed:
[Snyk] Upgrade recharts from 2.7.3 to 2.8.0 (#2950)

scudette committed:
Added FAT accessor for parsing FAT filesystems (#2961)

scudette committed:
[Snyk] Upgrade axios-retry from 3.6.0 to 3.6.1 (#2949)

scudette committed:
Implemented a writeback service to manage the writeback file. (#2966)

Added Registry based writeback

scudette committed:
[Snyk] Upgrade axios-retry from 3.6.1 to 3.7.0 (#2963)

hillu committed:
Add filesystem type to data returned by file accessor on Unix (#2967)

mgreen27 committed:
add suspicious field and targeted default (#2971)

- Add suspicious field and associated params
- Add specific default targeting

scudette committed:
[Snyk] Upgrade: @babel/core, @babel/plugin-transform-react-jsx, @babel/runtime (#2970)

mgreen27 committed:
Update Lnk.yaml (#2972)

Add some VQL optimisation.

jeffmahoney committed:
logscale: fix stats_interval parameter handling (#2973)

stats_interval describes seconds but is not being scaled up to seconds.

jeffmahoney committed:
Include MAC addresses in client host dashboard (#2943)

This PR extends the Generic.Client.Info artifact to add network
interface information to the output for Linux clients. Then, we can use
the mac address information for both Windows and Linux clients in the
host info dashboard.

---------

Co-authored-by: Mike Cohen <mike@velocidex.com>

scudette committed:
Implemented a fix for Windows.Timeline.Prefetch (#2974)

Fix: #2969

scudette committed:
Added a sample bash script for offline collector generation. (#2975)

mgreen27 committed:
Add Yara device scanning (#44) (#2978)

Add Yara device scanning and tests

scudette committed:
Fixed IPv6 formatting in Windows.Forensics.UserAccessLogs (#2980)

Fixes: #2968

scudette committed:
Various Bugfixes  (#2981)

Fix: #2947
Fix: #2946

scudette committed:
Update tool dependencies on each build (#2987) (#2989)

jeffmahoney committed:
utils/refcount: add simple refcount implementation

djoreilly commented:
I think the `SUSE: Add docker-compose environment` commit should be removed now that we are not using the kafka-humio-gateway anymore.

djoreilly commented:
We are missing a few cronsnoop fixes that were in 0.6.7.

djoreilly committed:
cronsnoop: fix panic when crontab has empty line

Blank lines are allowed in crontab files, but were failing with:
panic: runtime error: index out of range [0] with length 0

jeffmahoney commented:
> I think the `SUSE: Add docker-compose environment` commit should be removed now that we are not using the kafka-humio-gateway anymore.

No, because docker-compose is still the right way to deploy this in small environments.  It's simplified because we don't need the kafka-humio-gateway, but we still need the proxy. The commit also documents how to deploy.

jeffmahoney committed:
vql/linux/audit: actually add test cases

djoreilly commented on vql/linux/audit/audit_service_test.go:
```diff
@@ -0,0 +1,119 @@
+// +build linux
```
I had to add a blank line after this for the tests to run

jeffmahoney replied:
Fixed with https://github.com/SUSE/linux-security-sensor/pull/68/commits/1653021fd86fadb648860152634d305ce3f77503

djoreilly commented on vql/linux/audit/audit_service_test.go:
```diff
@@ -0,0 +1,119 @@
+// +build linux
+package audit
+
+import (
```
indents are a mix of tabs and spaces

jeffmahoney replied:
Fixed with https://github.com/SUSE/linux-security-sensor/pull/68/commits/1653021fd86fadb648860152634d305ce3f77503

djoreilly commented on vql/linux/audit/audit_mock_client_test.go:
```diff
@@ -1,3 +1,4 @@
+// +build linux
```
I had to add a blank line after this for the tests to run

jeffmahoney replied:
Fixed with https://github.com/SUSE/linux-security-sensor/pull/68/commits/1653021fd86fadb648860152634d305ce3f77503

djoreilly reviewed (changes requested):
also staticcheck found 
```
$ ~/go/bin/staticcheck vql/linux/audit/*go
vql/linux/audit/audit_listener.go:19:2: field timeout is unused (U1000)
vql/linux/audit/audit_mock_client_test.go:81:2: field client is unused (U1000)
vql/linux/audit/audit_mock_client_test.go:82:2: field scope is unused (U1000)
vql/linux/audit/audit_service.go:141:22: should use make([]byte, bufSize) instead (S1019)
vql/linux/audit/audit_service.go:162:6: type auditBufs is unused (U1000)
vql/linux/audit/audit_service.go:543:4: ineffective break statement. Did you mean to break out of the outer loop? (SA4011)
vql/linux/audit/audit_service.go:794:4: ineffective break statement. Did you mean to break out of the outer loop? (SA4011)

```

djoreilly commented on vql/linux/audit/audit_service.go:
```diff
@@ -0,0 +1,911 @@
+		if err != nil {
+			self.rawBufPool.Put(buf)
+			if errors.Is(err, syscall.EAGAIN) {
+				continue
```
This is causing ~100% cpu. Changing it to break seems to fix it.

jeffmahoney replied:
Fixed with https://github.com/SUSE/linux-security-sensor/pull/68/commits/d1d63c2764508f151e074602a1325e79c6dcf28b

jeffmahoney committed:
vql/linux/audit: don't overload EAGAIN

syscall.EAGAIN is returned by recv() if it's in non-blocking mode and
it would block (if no data is waiting), in which case we should go back
to waiting via epoll.  We also used it to indicate that the receive operation
should be retried.  The latter involved staying in the receive loop,
which is an obvious conflict.

This commit introduces a errRetryNeeded error that is used to indicate
that the operation should be retried and exits the loop on EAGAIN.

jeffmahoney committed:
vql/linux/audit: gofumpt -extra

jeffmahoney committed:
vql/linux/audit: fix staticcheck checks

jeffmahoney commented:
> also staticcheck found
> 
> ```
> $ ~/go/bin/staticcheck vql/linux/audit/*go
> vql/linux/audit/audit_listener.go:19:2: field timeout is unused (U1000)
> vql/linux/audit/audit_mock_client_test.go:81:2: field client is unused (U1000)
> vql/linux/audit/audit_mock_client_test.go:82:2: field scope is unused (U1000)
> vql/linux/audit/audit_service.go:141:22: should use make([]byte, bufSize) instead (S1019)
> vql/linux/audit/audit_service.go:162:6: type auditBufs is unused (U1000)
> vql/linux/audit/audit_service.go:543:4: ineffective break statement. Did you mean to break out of the outer loop? (SA4011)
> vql/linux/audit/audit_service.go:794:4: ineffective break statement. Did you mean to break out of the outer loop? (SA4011)
> ```

Fixed with https://github.com/SUSE/linux-security-sensor/pull/68/commits/765dab7f9d358519b7d382ab067a019466e62259

djoreilly reviewed (approved):
lgtm - thanks!
